text/2018-11-12-unified-key-format.md

```diff
@@ -0,0 +1,167 @@
+
+- Go format, like `t\x00\x00\x00\x00\x00\x00\x00\x1c_r\x00\x00\x00\x00\x00\x00\x00\xfa`
+
+Which creates the following problems for us:
+
+- Bad interoperability: These different formats are pain points when working with different tools
+  together. For example, pd-ctl produces keys in Go format while tikv-ctl accepts keys in
+  Protobuffer format.
+
+- Hard to implement: Actually there are some attempts to support different kind of formats, like
+  https://github.com/tikv/tikv/pull/3182 and https://github.com/pingcap/pd/pull/474. However it
+  turns out that these implementations won't work in real world scenario. For example, it is
+  not sufficient to decode the Golang format by only dealing with hex escapes `\x..` [(1)] and
+  decode the Protobuffer format by only dealing with oct escapes `\...` [(2)].
+
+- Hard to recognize different parts manually: For hex format, it is hard to extract different parts
+  from it manually.
+
+- Hard to be used in shells: For Protobuffer format and Go format, because there are escape
+  characters U+005C (\\) in the output, we need to do an escape (convert `\` to `\\`) when passing
+  it as a command line parameter in shells. As a result, current documentations are suggesting
+  wrong usages [(3)] that *happens* to work under specially crafted code, or won't work [(4)]
+  actually.
+
+- Contains ambiguous multi-bytes: For Go format, it directly outputs Unicode character if there is
+  a valid unicode sequence. For example, hex `00E6B58B01FF` will be printed as `\x00测\x01\xff` in
+  Go format. This may result in ambiguous bytes represents in different environments when such
+  characters are printed.
+
+To overcome above issues, we need a new human readable format and use it among all TiDB software
+stack components, which provides these feature:
+
+- Easy to recognize different parts after encoding by human.
+
+- Easy to implement encoding and decoding. Better to have a lot of existing libraries.
+
+- No escape character U+005C (\\) after encoding so that it can be easily used in shells.
+
+- No multi-byte characters after encoding to avoid encoding issues.
+
+# Detailed Design
+
+This RFC proposes to use hex format like `74000000000000001C5F7200000000000000FA` as unified a human
+readable format for keys:
+
+- Each byte must be encoded to a 2 digit hex in upper case.
+
+- Encoded hex string must be concatenated together without any other characters.
+
+- Decoder must support hex string in either lower-case or upper-case.
+
+- Decoder must throw errors when meeting non-hex or odd number of characters.
+
+The encoding process can be described by the JavaScript code below:
+
+```js
+function encode(buffer) {
+  return [...buffer]
+    .map(eachByte => eachByte.toString(16)) // convert to hex
+    .map(hex => hex.padStart(2, '0'))       // pad with zero
+    .map(hex => hex.toUpperCase())          // convert to upper case
+    .join('');
+}
+
+// encode(Buffer.from('Hello World'))
```

Comment:
(Not required) Might be worth showing a chinese and emoji including demo as well, to emphasize the compatibility with all inputs.

Reply:
Good idea